### ISSUE TITLE
Causal Loop Diagrams

### ISSUE THREAD

**Issue by bbak** [mermaid-js/mermaid#4453]
### Proposal

Please add [Causal Loop Diagrams](https://en.wikipedia.org/wiki/Causal_loop_diagram), because they are a preferred and well-known method, invented in the field of System Dynamics, to visualize and explore systems of causal relations between variables that may change over time.

As compared to the Flowchart Type, these would be the needed changes:

- Edges should always be shaped like a section of a circle (or loop).
- Positioning of nodes doesn't really have a direction like TD or similar; it could be more like inside-out or like Nodes at the top of the sourcecode have a tendency towards the center
- Any edge must denote the type of relationship between two nodes: either via a label like S, O or +, - to indicate if the edge represents a positive (Same direction) or negative (Opposite direction) relationship or by utilizing different edge styles.
- Ability to label loops: a variant of subsections may suit that.
- Ability to indicate a delay on an edge.
- The positioning of the nodes, and therefore length of the edges is quite important since the latter is an indicator for the duration of the (feedback-)loop.

Here are some resources regarding the construction of CLDs:

- [Guidelines for Drawing Causal Loop Diagrams](https://www.cs.toronto.edu/~sme/SystemsThinking/GuidelinesforDrawingCausalLoopDiagrams.pdf)
- The chapter on CLD in [Systems Mapping - How to build and use causal models of systems](https://link.springer.com/book/10.1007/978-3-031-01919-7) (open access PDF or E-Pub from Springer)
- A huge and quite famous example from the UK Government, the [Obesity System Map](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/296290/obesity-map-full-hi-res.pdf). That one uses different edge styles to distinguish S/O relationships. Additionally, I assume due to the size of the map, they didn't label the loops.

### Use Cases

Well, any type of dynamic system can be visualized by CLD's hence there's a multitude of use cases across a multitude of fields. 

### Screenshots

Here's a CLD I've done myself using [Kumu.io](https://kumu.io/bbak/wip)
<img width="694" alt="image" src="https://github.com/mermaid-js/mermaid/assets/4518127/ec305200-ce73-4bce-b58a-3e7f50de599a">

### Implementation

None

**Comment by abitrolly** [comment]
Some references from https://github.com/mermaid-js/mermaid/issues/177#issuecomment-1424478529 by @a-y-u-s-h:

>   1. [Introduction to Systems Thinking (by Lund University)](https://github.com/mermaid-js/mermaid/files/10699418/1.Introduction.to.Systems.Thinking.pdf)
> 
>   2. Existing Interactive Software for Causal Loop Diagrams - [Loopy, by Nicky Case](https://ncase.me/loopy/).
> 
>   3. [A Philophical look at System Dynamics](https://www.youtube.com/watch?v=XL_lOoomRTA&t=1400s)
> 
> 
> Influence Diagrams are closely related but they can be built with flowcharts or graphs.

**Comment by dudung** [comment]
Is it possible still using flowchart and with some elements styled using CSS, e.g. for reinforcing and balancing loop, time delayed, etc.?

**Comment by zcstarr** [comment]
Here's the work that I've done to get mermaid to support it , I came up with a template for doing CLD in mermaid. It's super useful to be able to share these documents as code.  The major blocker I think from full representation are Notes. 

https://github.com/mermaid-js/mermaid/issues/821#issuecomment-1852966539

Here's the style guide I'm trying to standardize with mermaidjs to be used in alot of places 
https://github.com/zcstarr/casually-looped

@abitrolly and @bbak  I'd be curious what you think? 

[Here's a more full mermaid example](https://mermaid.live/edit#pako:eNqFVF1v2kAQ_Cuni9IXDMWYUOxIlRBJJCQSEKQvNXm42ms4Ye7c8zmQAv-967MdcD5aP_l2Z3ZnZ-3b00CGQD26VCxZkfFsIQg-46k_5r8zHnL9QqZKPvMQVPpEms3vh_JIXgEHhNvneCnjp6LO7f3Uv91pUILF5J6pNegkZgFgTR5AUW8kojgDEUBe55VmUoMogkAfyPDRH0qRZhsU8fVRMSOmgA4fDdIEyY-Ui6XpbzRVs9gGcgeQki9kBlumwvTUrMqPNgmoDRMgNBnLFBGjsf82WHYdjUmzhZwZT9fkjgVaqlPBUtJ8yxKCfJZPMJ-OfBMwg5fhshjmCmNRd5kfrphYwgdDDNQvrhVbApkkiVQ6E1xzQK2Dif9JruwymBSSiy2QoVQKveVSHHK7F6JAXV6ifSnynoEwEZIHWDJzMPiY5YS0kmQUNc5UVntrvF9lI99izZ_GB-OdESuLm--MbRg76959UmwwqQ_frA0bxCxNbyAiIUQsizWJeBx7F5EbWalWcg3eheM45Xtzy0O98rrJ7voNO3798gu-G_U-53fe83XxQZfs3r-6f8CG3Un3f_uecdEmCz0uR7-uZ-zTULUMbqBUWwujpxZuwkKbUc41tegm_2t4iBfLPgcuqF7BBhbUw9ey5YIuxBGhLNNy_iIC6mmVgUWzJGQabjjDK2lTBRMmfkp5fqTenu6o57gtx7HtK9vtttvdvtPvWfSFerbb6tj9bsd1O1c9jPe_HS36x1Rot64sCjiaVPfF3WeuwONfM6aWHw)

In this one because of the lack of notes we just use lines in the same direction with labels that say what they are

**Comment by abitrolly** [comment]
@zcstarr it visually hard to compare flow chart diagram to circles used to depict casual loops. I also not sure if the format could be parsed by casual loop tools, which could be a nice feature for sharing and animating the dynamics.

**Comment by zcstarr** [comment]
> @zcstarr it visually hard to compare flow chart diagram to circles used to depict casual loops. I also not sure if the format could be parsed by casual loop tools, which could be a nice feature for sharing and animating the dynamics.

Not quite sure I understand, tbh I think animation is a different product. I've been thinking about using this as a description language and then going back and forth between a different system for say animating dynamics. 

I think nloopy really isn't that hard to parse both directions ( It uses a graph itself). Really maybe the issue is also that there's an opportunity to do more complex things. That said with in the realm of causal loop diagrams, I've never read or seen a formal spec that uses magnitude in the description. 

Perhaps there's an opportunity for us to make something more rigorous have you seen anyone set parameters for capturing that it in literature ? 

My problem with loopy and mermaid also comes from size. If your diagram gets complex the universe for drawing the diagram can get pretty chaotic. For mermaid left to right flow chart style can get long and there's lots of crossing lines , which may not be too avoidable.

**Comment by martinzokov** [comment]
I'd be really interested in having this as part of mermaid. Any chance of implementing this?

**Comment by BradKML** [comment]
Would like to keep an eye on this + conversion to something "fun" https://github.com/schucan/CLD

**Comment by dudung** [comment]
> Would like to keep an eye on this + conversion to something "fun" https://github.com/schucan/CLD

Very easy to use. Thanks. Is there anyway to embed it to a Hugo post or Medium story?

**Comment by victorzhrn** [comment]
For anyone who lands here wanting a causal loop diagram: mermaid doesn't do these, but I built a small standalone lib that does (disclosure: it's mine) — link polarities (`+`/`-`), delay marks on the slow links, and it auto-detects and labels the reinforcing (R) / balancing (B) loops:

![Causal loop diagram with R/B loops, link polarities and delay marks](https://schematex.js.org/og/render?d=Y2F1c2FsbG9vcCAiV2h5IHRoZSBoZWxwZGVzayBxdWV1ZSBuZXZlciBjbGVhcnMiCiAgIk9wZW4gdGlja2V0cyIgLT4gIkFnZW50IHN0cmVzcyIgOiArCiAgIkFnZW50IHN0cmVzcyIgLT4gIlJlc29sdXRpb24gc3BlZWQiIDogLQogICJSZXNvbHV0aW9uIHNwZWVkIiAtPiAiT3BlbiB0aWNrZXRzIiA6IC0KICAiT3BlbiB0aWNrZXRzIiAtPiAiT3ZlcnRpbWUgaG91cnMiIDogKwogICJPdmVydGltZSBob3VycyIgLT4gIlJlc29sdXRpb24gc3BlZWQiIDogKyBkZWxheQogIGxvb3AgQjEgIkJ1cm5vdXQgYmFsYW5jaW5nIgo)

```text
causalloop "Why the helpdesk queue never clears"
  "Open tickets" -> "Agent stress" : +
  "Agent stress" -> "Resolution speed" : -
  "Resolution speed" -> "Open tickets" : -
  "Open tickets" -> "Overtime hours" : +
  "Overtime hours" -> "Resolution speed" : + delay
  loop B1 "Burnout balancing"
```

Stopgap for now — would love to see a native version land. Docs: https://schematex.js.org/docs/causalloop

**Comment by volkanunsal** [comment]
@victorzhrn Great little language! I made a VSCode extension for it here.

https://marketplace.visualstudio.com/items?itemName=VolkanUnsal.schematex-vscode
